src/viewer/input.js

```diff
@@ -0,0 +1,107 @@
+import {
+  Raycaster,
+  LineBasicMaterial,
+  Line,
+  Matrix4,
+  BufferGeometry,
+  Vector3,
+} from "three";
+
+import Renderer from "../engine/renderer";
+import State from "../engine/state";
+
+class Input {
+  constructor(scene, ui) {
+    this.scene = scene;
+    this.ui = ui;
+    this.scene.traverse(child => {
+      if (child.name === "OrbContainer") {
+        this.orbContainer = child;
+        return;
+      }
+    });
+    // XR controllers
+    this._raycaster = new Raycaster();
+    this._tempMatrix = new Matrix4();
+    this._controller1 = Renderer.xr.getController(0);
+    this._controller2 = Renderer.xr.getController(1);
+    this._controller1.handedness = "left";
+    this._controller2.handedness = "right";
+    this._controllers = [this._controller1, this._controller2];
+    this._controllers.forEach(controller => {
+      this.scene.add(controller);
+      this.addRayCastVisualizer(controller);
+      controller.Update = () => {
+        controller._raycastAt = this.raycast(controller);
+        controller.children[0].material.opacity =
+          controller._raycastAt == undefined ? 0 : 1;
+      };
+    });
```

Comment:
This code should end up being much cleaner once controllers can be provided directly from Sandcastle's XRInput.

Reply:
as well as raycasting!